src/dmd/expressionsem.d

```diff
@@ -313,9 +313,34 @@ Expression resolveOpDollar(Scope* sc, ArrayExp ae, IntervalExp ie, Expression* p
         return e;
     }
 
+    int is_dollar = 0;
+
+    auto lwr = ie.lwr;
+    auto upr = ie.upr;
+
+    // check if lower or upper were dollar
+    if (lwr.isIdentifierExp())
+    {
+        if (lwr.isIdentifierExp().ident == Id.dollar)
+            is_dollar = 1;
+    }
+    if (upr.isIdentifierExp())
+    {
+        if (upr.isIdentifierExp().ident == Id.dollar)
+            is_dollar = 1;
+    }
+
     ie.lwr = sem(ie.lwr);
     ie.upr = sem(ie.upr);
 
+    // if there was an error and upper and lower were dollar,
+    // then I should give a supplemental error opover.d
+    // (after the call of this function)
+    if ((ie.lwr.isErrorExp() || ie.upr.isErrorExp()) && is_dollar == 1)
```

Comment:
`is_dollar` is not needed and the check is wrong anyway. You are checking only for simple uses of `$` such as `a[0..$]`. More complicated uses such as `a[$-7..$-2]` will not pass this check, but that is besides the point. The check at line 344 will declare a `$` variable if it was used. The temporary is returned via `pe0` otherwise it's null. This could be checked in opover.